src/mlpack/methods/ann/augmented/ham_unit_impl.hpp

```diff
@@ -0,0 +1,99 @@
+/**
+ * @file ham_unit_impl.hpp
+ * @author Konstantin Sidorov
+ *
+ * Implementation of HAMUnit class
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_AUGMENTED_HAM_UNIT_IMPL_HPP
+#define MLPACK_METHODS_AUGMENTED_HAM_UNIT_IMPL_HPP
+
+#include "ham_unit.hpp"
+
+namespace mlpack {
+namespace ann /* Artificial Neural Network */ {
+namespace augmented /* Augmented neural network */ {
+
+HAMUnit::HAMUnit(int memorySize,
+                 C& controller,
+                 E& embed,
+                 J& join,
+                 S& search,
+                 W& write)
+  : memorySize(memorySize),
+    search(search), embed(embed), controller(controller), t(0)
+{
+  memory = TreeMemory<double, J, W>(memorySize, join, write);
+}
+
+arma::vec HAMUnit::Attention() const
+{
+  size_t nodesCnt = 2 * memory.ActualMemorySize() - 1;
+  arma::vec probabilities(nodesCnt);
+  probabilities(0) = 1;
+  arma::vec hController = sequence.col(t++);
+  for (size_t node = 1; node < nodesCnt; ++node) {
+    size_t parent = memory.Parent(node);
+    bool dir = node == memory.Left(parent);
+    arma::vec h(hController.n_elem + memory.GetCell(parent).n_elem);
+    h.rows(0, hController.n_elem - 1) = hController;
+    h.rows(hController.n_elem,
+           hController.n_elem + memory.GetCell(parent).n_elem - 1)
+        = memory.GetCell(parent);
+    double prob = search.Forward(h);
+    if (!dir) prob = 1. - prob;
+    probabilities(node) = prob * probabilities(parent);
+  }
+  size_t from = mem.LeafIndex(0),
+         to = memory.LeafIndex(memory.MemorySize() - 1);
+  arma::vec leafAttention = probabilities.rows(from, to);
+  assert(abs(arma::accu(leafAttention) - 1) < 1e-4);
+  return leafAttention;
+}
+
+void HAMUnit::Forward(arma::mat&& input, arma::mat&& output) {
+  sequence = input;
+  memory.Initialize(sequence);
+  t = 0;
+
+  arma::vec attention = Attention();
+  arma::vec input = arma::zeros(memory.Get(0).n_elem);
+  for (size_t i = 0; i < memorySize; ++i)
+  {
+    input += leafAttention.at(i) * memory.Get(i);
+  }
+  // TODO Feed input to the controller and get output from it.
+
+  // Write phase.
+  for (size_t i = 0; i < memory.MemorySize(); ++i)
+  {
+    memory.Get(i) *= attention.at(i);
+    memory.Get(i) += (1. - attention.at(i)) * write.Forward(sequence.col(t++));
```

Comment:
I think WRITE takes `attention` and `memory.Get(i)` as input parameters, is that correct?  Maybe I have overlooked something and this code is equivalent to that part of the paper, I am not sure.